New fields in account keys and public addresses

garbageslam opened the pull request:
Soundtrack of this PR: https://www.youtube.com/watch?v=gHUBYt1oPio

### Motivation

We must make some changes where we add some fields to account key and public address.
We need to incorporate the fog authority key fingerprint and the user's signature over that fingerprint,
in order to establish a complete chain of trust back to Bob whenever Alice wants to use a fog report
to send money to Bob.
In earlier discussion we considered encoding this as query parameters in fog-url, but that was only
to avoid making any changes to the b58-encoding. The b58-encoding has changed a lot since then
and has become much simpler to update.
We also should add version numbers to account keys / public addresses soon if we want to have that
for release. Otherwise it will be difficult or impossible to change the meaning of any of the fields. 

### In this PR

 Add fog_authority_key_fingerprint to AccountKey, sig to PublicAddress

This just adds fields to the structures, it doesn't actually compute
signatures or give them the right values, or check the signatures.

It also adds version numbers to both structs. This gives us an
escape if we have to make backwards incompatible changes to the way
that public addresses are handled, fog reports are looked up, change
the interpretations of some of these fields, etc.

This also removes Serde serialization of public address, which was
only being used in the keyfile. It makes public address implement
prost instead.

This also removes external.RistrettoPublic from the proto for
public address. It replaces it with CompressedRistretto and adds
some more missing conversions. This leads to small changes in mobilecoind.

This also updates the b58-payloads format to include the version number
and the fog authority sig. I believe that this change is backwards
compatible for any payload not containing fog data.

We should possible delete RistrettoPublic message in external.proto,
I believe that it should never be used.

### Future Work

- Remove RistrettoPublic from external.proto
- Remove serialization traits for RistrettoPublic rust type
- Consider making a simpler URL encoding of Public Address using query parameters to encode different optional components. - While the b58 encoding is very small on the wire, it is hard to create without having libmobilecoin. A version using query parameters may be easier for a human to inspect, and easy to create and destructure in javascript even without having libmobilecoin, i.e. having only a standard url lib and protobufs. (this is https://github.com/mobilecoinofficial/mobilecoin/pull/242)

### TODO

Try to ensure that this will not break testnet, or accept that users may have to download mobilecoind again, and learn from what happened.

mfaulk commented on transaction/core/src/account_keys.rs:
```diff
@@ -71,30 +93,41 @@ impl PublicAddress {
     /// `spend_public_key` - The user's public subaddress spend key `D`,
     /// `view_public_key` - The user's public subaddress view key `C`,
     /// `fog_url` - User's fog url
+    /// `
```
For consistency, can you add doc comments for the new args?

kylefleming commented on transaction/core/src/account_keys.rs:
```diff
@@ -71,30 +93,41 @@ impl PublicAddress {
     /// `spend_public_key` - The user's public subaddress spend key `D`,
     /// `view_public_key` - The user's public subaddress view key  `C`,
```
@mfaulk's comment applies here as well.

kylefleming commented on transaction/core/src/account_keys.rs:
```diff
@@ -110,26 +143,38 @@ impl PublicAddress {
-    spend_private_key: RistrettoPrivate,
+    pub spend_private_key: RistrettoPrivate,
 
     /// Fog URL (if user has Fog service)
```
Could you add to the comment to mention what the value will be if the user doesn't have a Fog service?

kylefleming commented on transaction/core/src/account_keys.rs:
```diff
@@ -110,26 +143,38 @@ impl PublicAddress {
-    fog_url: String,
+    pub fog_url: String,
+
+    /// Fog Authority Key Fingerprint (if user has Fog service)
```
Same here. What is the value if the user doesn't have a Fog service?

kylefleming commented on transaction/core/src/account_keys.rs:
```diff
@@ -13,37 +13,59 @@
+    /// A signature with the user's view_private_key over the fog authority key fingerprint.
+    /// Empty if no fog for this address
+    #[prost(bytes, tag = "4")]
+    pub fog_authority_sig: Vec<u8>,
```
Could you add an accessor that returns `Option<&[u8]>` in the same vein as `fog_url(&self) -> Option<&str>`?

kylefleming commented on transaction/core/src/account_keys.rs:
```diff
@@ -110,26 +143,38 @@ impl PublicAddress {
+
+    /// Fog Authority Key Fingerprint (if user has Fog service)
+    #[prost(bytes, tag = "4")]
+    pub fog_authority_key_fingerprint: Vec<u8>,
```
Same here, could you add an accessor that returns `Option`?

kylefleming commented on transaction/core/src/account_keys.rs:
```diff
@@ -261,14 +319,20 @@ impl AccountKey {
+            version: self.version,
+        };
+
+        if !self.fog_url.is_empty() {
```
Style-wise, this could use the Option accessor to hide the details of an empty string representing the absense of a Fog url.

garbageslam replied:
I think I see what you are saying but I don't know who we are hiding the details from, this is all implementation details of the same object

kylefleming replied:
Not in terms of hiding from something, but just in terms of having the logic go through the same code path w.r.t. whether the AccountKey is considered to use Fog or not, so that it's less likely to crop up as subtle bugs later if that logic were to change (e.g. empty string no longer signifies that the AccountKey doesn't use Fog).

I agree that since it's an implementation detail of the same object, it's less of an issue. Feel free to ignore this PR comment though if you prefer how it is.

kylefleming commented on util/b58-payloads/src/payloads.rs:
```diff
@@ -101,9 +101,15 @@ pub struct RequestPayload {
     /// The spend public key.
     pub spend_public_key: [u8; 32],
 
+    /// The public address version
```
We should make sure @rjwalters also reviews the changes to `RequestPayload`.

kylefleming commented on transaction/core/src/account_keys.rs:
```diff
@@ -13,37 +13,59 @@
-    fog_url: Option<String>,
+    /// Should be parseable as mc_util_uri::FogUri.
+    #[prost(string, tag = "3")]
+    pub fog_url: String,
```
This should probably also have a comment regarding empty

eranrund commented:
Repeating my comment from https://github.com/mobilecoinofficial/mobilecoin/pull/242:
I started going over this but I think we really need to rethink the version field.
In its current incarnation it adds a lot of overhead for, from what I can tell, no value.
There are 2 objects: PublicAddress (rust) and it's proto representation. For the proto side we could either have different proto definitions (PublicAddressV1, etc), or rely on proto's forward-compatibility via tag numbers. For the Rust representation, this feels to me like the Transaction case in which we can make appropriate changes when the time arises that we need to change the internal representation.

eranrund commented:
Also please see other comments in #242 regarding plaguing the constructor with the version field everywhere.

garbageslam commented:
@kylefleming @eranrund i removed the version field, let me know if this looks alright to you

@rjwalters I added additional validation of the fog-url field in the b58-payloads. I think that will be a breaking change for which payloads are accepted but I think it is necessary, lmk what you think

garbageslam commented:
@sugargoat btw this is going to change the format of the keyfiles because they don't use serde-json now. AFAIK that doesn't cause an issue, its only in the ops data containers I think?

eranrund commented on api/proto/external.proto:
```diff
@@ -53,31 +53,54 @@ message Ed25519Signature {
 
-    // Optional url of fog server. Empty string when not in use.
+    /// Optional url of fog report server.
+    /// Empty string when not in use, i.e. for accounts that don't have fog service.
```
If this is the report server, can we call it `fog_report_url`?

garbageslam replied:
yah i'm on board with that

eranrund commented on transaction/core/src/account_keys.rs:
```diff
@@ -108,30 +139,64 @@ impl PublicAddress {
     /// Private key 'a' used for view-key matching.
     #[prost(message, required, tag = "1")]
-    view_private_key: RistrettoPrivate,
+    pub view_private_key: RistrettoPrivate,
```
Do these need to be `pub`? Ideally nothing  in this struct should be `pub`, and there should be accessors and constructors.

garbageslam replied:
ok -- i guess i started taking the point of view that, `AccountKey` is just plain-old-data and there is no invariant here to protect.

I mean really any values of these members is correct. Any `RistrettoPrivate` value can be the `view_private_key` etc. Any url can be the `fog_url`. Any bytes can be the `fog_authority_key_fingerprint`. Any string can be the `fog_report_key`.

Also like, if there were some invariant that must be enforced in the constructor, it would be bypassed by `prost::Message`.

So stylistically I would say, these should all be pub, and it's okay to treat it as plain old data.

Maybe for public address that's not really the case because the `sig` has some dependency on the keys, and the values would not be correct if the keys or sig changed.

garbageslam replied:
I will try removing the `pub` stuff and see what happens, I had thought that `prost::Message` requires the members to be `pub` at one point, but I guess that actually can't be the case

garbageslam replied:
i think that worked out fine, lmk what you think

garbageslam commented on mobilecoind/src/service.rs:
```diff
@@ -376,12 +376,13 @@ impl<T: BlockchainConnection + UserTxConnection + 'static> ServiceApi<T> {
-            fog_url,
+            receiver.fog_report_url().unwrap_or(&""),
+            receiver.fog_authority_sig().unwrap_or(&[]),
+            receiver.fog_report_key().unwrap_or(&""),
```
@eranrund lmk if you think this pattern is okay

eranrund replied:
LGTM

garbageslam commented on util/b58-payloads/src/payloads.rs:
```diff
@@ -71,11 +76,15 @@ fn decode_payload(
+    // Empty fog url is okay, otherwise it must parse as FogUri
+    // TODO: Reject uri's containing a path?
+    if !fog_report_url.is_empty() {
+        FogUri::from_str(fog_report_url).map_err(Error::FogUrlParsingError)?;
```
@rjwalters @kylefleming @voloshyn this may cause old payloads not to be valid anymore, if the fog urls dont have `fog://` or `insecure-fog://` scheme

garbageslam commented on mobilecoind-json/src/main.rs:
```diff
@@ -244,9 +244,48 @@ fn request_code(
+}
+
+// Helper conversion between json and protobuf
+impl TryFrom<&JsonPublicAddress> for PublicAddress {
```
this addresses comment https://github.com/mobilecoinofficial/mobilecoin/pull/242#discussion_r446309937

eranrund commented on api/proto/external.proto:
```diff
@@ -53,31 +53,54 @@ message Ed25519Signature {
+    /// Optional fog report key.
+    /// The fog report server may serve multiple reports, this key disambiguates
+    /// which one to use when sending to this account.
+    string fog_report_key = 5;
```
I think it'll be nicer if this is next to fog_report_url.

eranrund commented on util/b58-payloads/src/payloads.rs:
```diff
@@ -1,11 +1,16 @@
 use mc_transaction_core::account_keys::{AccountKey, PublicAddress};
 use mc_transaction_std::identity::RootIdentity;
+use mc_util_uri::FogUri;
+
```
Why did this move here? In most places we just group all use statements and let rustfmt order them.

garbageslam replied:
james has a pattern where he separates like `core/std` from `mc` from 3rd party libs, so i put crc at the bottom

the FogUri dep is new, I'm adding it to help validate the `fog_report_url` values below

jcape replied:
I group them in the cargo.toml, because "path" crates are special, but I just bunch all the "use" statements together as @eranrund is suggesting. The only ones I put elsewhere are things like `alloc::vec` where they get improperly merged...

garbageslam replied:
i see i got confused

eranrund commented on util/keyfile/src/lib.rs:
```diff
@@ -37,7 +37,7 @@ pub fn read_keyfile_data<R: std::io::Read>(buffer: &mut R) -> Result<RootIdentit
 /// Write user public address to disk
 pub fn write_pubfile<P: AsRef<Path>>(path: P, addr: &PublicAddress) -> Result<(), std::io::Error> {
-    File::create(path)?.write_all(&serde_json::to_vec(&addr)?)?;
+    File::create(path)?.write_all(&mc_util_serial::encode(addr))?;
```
Maybe this breaks some Python code somewhere?

garbageslam replied:
hmm... okay, lets talk to @rjwalters
